concepts/websockets.textile

```diff
@@ -0,0 +1,593 @@
+
+You may have heard WebSockets simultaneously referred to both as a "transport" and as a "protocol". The former is more correct, because while they are a protocol in the sense that a strict set of rules for establishing communication and enveloping the transmitted data must be adhered to, the standard does not take any stance regarding how the actual data payload is structured within the outer message envelope. In fact, part of the specification includes the option for the client and server to agree on a protocol with which the transmitted data will be formatted and interpreted. The standard actually refers to these as "subprotocols", in order to avoid issues of ambiguity in the nomenclature. Examples of subprotocols are JSON, XML, MQTT, WAMP, et al, and these can ensure agreement not only about the way the data is structured, but also the way communication must commence, continue and eventually terminate. As long as both parties understand what the protocol entails, anything goes. The WebSocket simply provides a transport layer over which that messaging process can be implemented, which is why most common subprotocols are not exclusive to WebSocket-based communications.
+
+h4(#a-quick-note-about-authentication-and-authorization). A quick note about authentication and authorization
+
+Seeing as WebSockets is a thin layer built on top of TCP/IP, anything beyond the basic handshake and specification for message framing is really something that needs to be handled either on a per-application or per-library basis. "Quoting the RFC":https://tools.ietf.org/html/rfc6455#section-10.5:
+
+bq. _This protocol doesn't prescribe any particular way that servers can authenticate clients during the WebSocket handshake. The WebSocket server can use any client authentication mechanism available to a generic HTTP server, such as cookies, HTTP authentication, or TLS authentication._
+
+
+
+In a nutshell, use the HTTP-based authentication methods you'd use anyway, or use a subprotocol such as "MQTT":http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html or "WAMP":https://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.html, both of which offer approaches for authentication and authorization.
+
+h3(#getting-the-ball-rolling-with-http). Getting the ball rolling with HTTP
+
+One of the early considerations when defining the WebSocket standard was to ensure that it "play nicely" with the web. This meant recognising that the web is generally addressed using URLs, not IP addresses and port numbers, and that a WebSocket connection should be able to take place with the same initial HTTP-based handshake used for any other type of web request.
```

Comment:
'recognising' -> 'recognizing' (We use American English as per brand guidelines, let me know if you don't have access to it)